transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala

```diff
@@ -279,34 +295,44 @@ object HcaPipelineBuilder extends PipelineBuilder[Args] {
           case Some(schema) => {
             // if the schema is not able to load, throw an exception, otherwise try to use it to validate
             Schema.loadFromString(schema) match {
-              case Failure(_) =>
-                throw new Exception(
-                  s"Schema not loaded properly for schema at $url, file $filename"
-                )
+              case Failure(_) => {
+                val errorMessage = s"Schema not loaded properly for schema at $url, file $filename"
+                logSchemaValidationError(filename, errorMessage)
+                throw new Exception(errorMessage)
+              }
               case Success(value) =>
                 // try to parse the actual data into a json format for validation
                 parse(encode(data)) match {
-                  case Left(_) =>
-                    throw new Exception(s"Unable to parse data into json for file $filename")
+                  case Left(_) => {
+                    val errorMessage = s"Unable to parse data into json for file $filename"
+                    logSchemaValidationError(filename, errorMessage)
+                    throw new Exception(errorMessage)
+                  }
                   // if everything is parsed/encoded/etc correctly, actually try to validate against schema here
                   // if not valid, will return list of issues
                   case Right(success) => {
                     value.validate(success) match {
                       case Validated.Valid(_) => (filename, data)
-                      case Validated.Invalid(e) =>
-                        throw new Exception(
-                          s"Data does not conform to schema: ${e.map(_.getMessage)}"
-                        )
+                      case Validated.Invalid(e) => {
+                        val errorMessage =
+                          s"Data does not conform to schema from $url; ${e.map(_.getMessage).toList.mkString(",")}"
+                        logSchemaValidationError(filename, errorMessage)
+                        throw new Exception(errorMessage)
+                      }
                     }
                   }
                 }
             }
           }
-          case None => throw new Exception(s"No schema found at $url for file $filename")
+          case None => {
+            val errorMessage = s"No schema found at $url for file $filename"
+            logSchemaValidationError(filename, errorMessage)
+            throw new Exception(errorMessage)
+          }
         }
       }
     }
-    filenamesAndMsg
+    filenameAndMsg
```

Comment:
Since this code was modified to not side effect its arguments, it might be clearer to make this method `void` so it's explicit that this method is called purely for its side effects. That would avoid the need to return any value here and you could change
```
    val content = filenameAndMsg.map {
```
to use `for`

Reply:
Unfortunately, `SCollection`s don't have a `foreach` method, since they are the collection type used for Beam and Beam strictly doesn't allow side effects

Reply:
The SCollection and scio stuff is new to me. It's a bit odd but you're right, no foreach..